tests/test_ha.py

```diff
@@ -643,12 +643,40 @@ def test_manual_failover_process_no_leader(self):
         # same as previous, but set the current member to nofailover. In no case it should be elected as a leader
         self.ha.patroni.nofailover = True
         self.assertEqual(self.ha.run_cycle(), 'following a different leader because I am not allowed to promote')
+
+    def test_manual_failover_process_no_leader_in_synchronous_mode(self):
         # in sync mode only the sync node is allowed to take over
-        self.ha.cluster = get_cluster_initialized_without_leader(failover=Failover(0, 'leader', 'other', None))
-        self.ha.patroni.nofailover = False
+        self.p.is_leader = false
         self.ha.is_synchronous_mode = true
+
+        # switchover to a specific node, which name doesn't match our name (postgresql0)
+        self.ha.cluster = get_cluster_initialized_without_leader(failover=Failover(0, 'leader', 'other', None))
+        self.assertEqual(self.ha.run_cycle(), 'following a different leader because i am not the healthiest node')
+
+        # switchover to from a specific leader, but our name (postgresql0) in not in the sync nodes list
+        self.ha.cluster = get_cluster_initialized_without_leader(failover=Failover(0, 'leader', '', None),
+                                                                 sync=('leader', 'blabla'))
         self.assertEqual(self.ha.run_cycle(), 'following a different leader because i am not the healthiest node')
 
+        # our name (postgresql0) isn't in the /sync key and the `other` node is not available
+        self.ha.fetch_node_status = get_node_status(nofailover=True)  # accessible, in_recovery
+        self.ha.cluster = get_cluster_initialized_without_leader(failover=Failover(0, '', 'other', None),
+                                                                 sync=('leader1', 'blabla'))
+        self.assertEqual(self.ha.run_cycle(), 'following a different leader because i am not the healthiest node')
+
+        # the `other` node isn't available but our name is in the /sync key
+        self.ha.cluster = get_cluster_initialized_without_leader(failover=Failover(0, '', 'other', None),
+                                                                 sync=('leader1', 'postgresql0'))
+        self.p.pick_synchronous_standby = Mock(return_value=([], []))
+        self.ha.dcs.write_sync_state = true
+        self.assertEqual(self.ha.run_cycle(), 'promoted self to leader by acquiring session lock')
+
+        # the manual failover to the unsync node (the leader and all sync nodes are not available)
```

Comment:
This is an impossible scenario: https://github.com/zalando/patroni/blob/88db6018ac65afa9231f36f0d3c135c24a16e350/patroni/ha.py#L954-L957

The leader will refuse to demote if the candidate isn't synchronous.

Reply:
So could you please explain, following the logic of the tests implemented here - what is the expected behaviour of this? Or it is absolutely impossible to get this picture?
```
self.ha.cluster = get_cluster_initialized_without_leader(failover=Failover(0, '', 'postgresql0', None), 
                                                         sync=('leader', 'blabla'))
self.assertEqual(self.ha.run_cycle(), ?)
```

I am a bit confused because the code you refer is actually for `process_manual_failover_from_leader`, not `manual_failover_process_no_leader`

Reply:
Ok. In fact, it could happen that there will be a valid sync node(s), but they refuse to take the lock and promote because of the `nofailover` tag.
The operator in this case may initiate a manual failover to the async node.
Since the failover candidate may also disappear, other nodes in the Patroni cluster checking whether the candidate is healthy, and in case it is not they trigger the normal leader race. For a sync mode, only members listed in the /sync key may participate in the race, other nodes should report 'following a different leader because i am not the healthiest node'.